topics/engineering/improving-wiki-rif-search-in-genenetwork.gmi

```diff
@@ -0,0 +1,164 @@
+    Just("ProbeSetFreeze.confidentiality < 1 AND ProbeSetFreeze.public > 0"))
+```
+
+Simply doing a simple count for the above query takes 4 minutes 3.73 seconds.  The count query is:
+
+```
+SELECT COUNT(*) FROM Species
+INNER JOIN InbredSet ON InbredSet.SpeciesId = Species.Id
+INNER JOIN ProbeFreeze ON ProbeFreeze.InbredSetId = InbredSet.Id
+INNER JOIN Tissue ON ProbeFreeze.TissueId = Tissue.Id
+INNER JOIN ProbeSetFreeze
+ON ProbeSetFreeze.ProbeFreezeId = ProbeFreeze.Id
+INNER JOIN ProbeSetXRef ON
+ProbeSetXRef.ProbeSetFreezeId = ProbeSetFreeze.Id
+INNER JOIN ProbeSet
+ON ProbeSet.Id = ProbeSetXRef.ProbeSetId
+LEFT JOIN Geno ON ProbeSetXRef.Locus = Geno.Name
+AND Geno.SpeciesId = Species.Id;
+```
+
+For each row returned, we index each field I.e. description, symbol, etc as such:
+
+
+```
+index_text_without_positions = lambda text: termgenerator.index_text_without_positions(text)
+index_authors = lambda authors: termgenerator.index_text(authors, 0, "A")
+index_species = lambda species: termgenerator.index_text_without_positions(species, 0, "XS")
+index_group = lambda group: termgenerator.index_text_without_positions(group, 0, "XG")
+index_tissue = lambda tissue: termgenerator.index_text(tissue, 0, "XI")
+index_dataset = lambda dataset: termgenerator.index_text(dataset, 0, "XDS")
+index_symbol = lambda symbol: termgenerator.index_text_without_positions(symbol, 0, "XY")
+index_chr = lambda chr: termgenerator.index_text_without_positions(chr, 0, "XC")
+index_peakchr = lambda peakchr: termgenerator.index_text_without_positions(peakchr, 0, "XPC")
+
+[...]
+
+trait["mean"].bind(partial(add_mean, doc))
+trait["lrs"].bind(partial(add_peak, doc))
+trait["geno_mb"].bind(partial(add_peakmb, doc))
+trait["additive"].bind(partial(add_additive, doc))
+trait["year"].bind(partial(add_year, doc))
+
+# Index free text.
+for key in ["description", "authors", "dataset"]:
+    trait[key].bind(index_text)
+for key in ["Abstract", "Title"]:
+    trait.pop(key).bind(index_text)
+for key in ["species", "group", "inbredsetcode"]:
+    trait[key].bind(index_text_without_positions)
+for key in ["abbreviation", "Lab_code"]:
+    trait.pop(key).bind(index_text_without_positions)
+
+# Index text with prefixes.
+trait["species"].bind(index_species)
+trait["group"].bind(index_group)
+trait["authors"].bind(index_authors)
+trait["geno_chr"].bind(index_peakchr)
+trait["dataset"].bind(index_dataset)
+
+[...]
+```
+
+With the above indexing, even without RIF+WIKI indices, we take around 1 hour 15 minutes to index.  To add generif+wiki indexes, one naive solution would be to add an extra join to the relevant tables, in our case: "GeneRIF" (WIKI) and "GeneRIF_BASIC" (RIF).   However, the joins take time, since we'd have to do a join a VARCHAR column, "symbol", and this is inefficient.  Simply adding a join to GeneRIF_BASIC table takes XXXX minutes, and balloons the numbers of rows to XXXX rows:
```

Comment:
I don't think joins are slow because we use a VARCHAR column. Most probably it's because we haven't indexed this column. In this case the `Probeset` table doesn't have this column indexed.

```
SHOW INDEXES FROM `ProbeSet`

```

Also I think showing and explaining the output of explain would provide better context for why this is slow.
```
+------+-------------+----------------+--------+---------------------------------------+------------+---------+-----------------------------------+-------+--------------------------+
| id   | select_type | table          | type   | possible_keys                         | key        | key_len | ref                               | rows  | Extra                    |
+------+-------------+----------------+--------+---------------------------------------+------------+---------+-----------------------------------+-------+--------------------------+
|    1 | SIMPLE      | GeneRIF_BASIC  | index  | NULL                                  | symbol     | 774     | NULL                              | 98119 | Using index              |
|    1 | SIMPLE      | ProbeSet       | ref    | PRIMARY,symbol_IDX,ft_ProbeSet_Symbol | symbol_IDX | 403     | func                              | 1     | Using where; Using index |
|    1 | SIMPLE      | ProbeSetXRef   | ref    | ProbeSetFreezeId,ProbeSetId           | ProbeSetId | 4       | gn2.ProbeSet.Id                   | 2     |                          |
|    1 | SIMPLE      | ProbeSetFreeze | eq_ref | PRIMARY                               | PRIMARY    | 2       | gn2.ProbeSetXRef.ProbeSetFreezeId | 1     |                          |
|    1 | SIMPLE      | ProbeFreeze    | eq_ref | PRIMARY                               | PRIMARY    | 2       | gn2.ProbeSetFreeze.ProbeFreezeId  | 1     | Using where              |
|    1 | SIMPLE      | InbredSet      | eq_ref | PRIMARY                               | PRIMARY    | 2       | gn2.ProbeFreeze.InbredSetId       | 1     | Using where              |
|    1 | SIMPLE      | Species        | eq_ref | PRIMARY                               | PRIMARY    | 2       | gn2.InbredSet.SpeciesId           | 1     | Using index              |
|    1 | SIMPLE      | Tissue         | eq_ref | PRIMARY                               | PRIMARY    | 2       | gn2.ProbeFreeze.TissueId          | 1     | Using index              |
+------+-------------+----------------+--------+---------------------------------------+------------+---------+-----------------------------------+-------+--------------------------+

```